### PR TITLE
Add LLVM level allocation optimization pass

### PR DESCRIPTION
**Pull request by yuyichao**
Test might be added later.

Just to give a taste of it though

```julia
julia> function test_opt(a::Float64)
           s = Ref{Float64}()
           c = Ref{Float64}()
           ccall((:sincos, Base.libm_name), Void, (Float64, Ptr{Float64}, Ptr{Float64}), a, s, c)
           s[], c[]
       end
test_opt (generic function with 1 method)

julia> for i in 1
           @time test_opt(1.0)
       end
  0.000005 seconds

julia> @code_llvm test_opt(1.0)
```
```llvm
define void @julia_test_opt_63022([2 x double] addrspace(11)* noalias nocapture sret, double) #0 !dbg !5 {
top:
  %2 = alloca i64, align 8
  %3 = bitcast i64* %2 to %jl_value_t*
  %4 = alloca [2 x double], align 8
  %5 = bitcast %jl_value_t* %3 to double*
  %6 = getelementptr inbounds [2 x double], [2 x double]* %4, i64 0, i64 0
  call void inttoptr (i64 140684203667904 to void (double, double*, double*)*)(double %1, double* %6, double* %5)
  %7 = load i64, i64* %2, align 16
  %8 = getelementptr inbounds [2 x double], [2 x double]* %4, i64 0, i64 1
  %9 = bitcast double* %8 to i64*
  store i64 %7, i64* %9, align 8
  %10 = bitcast [2 x double]* %4 to i8*
  %11 = bitcast [2 x double] addrspace(11)* %0 to i8 addrspace(11)*
  call void @llvm.memcpy.p11i8.p0i8.i32(i8 addrspace(11)* %11, i8* %10, i32 16, i32 8, i1 false)
  ret void
}
```
```julia
julia> function f(a)
           b = Ref(a)
           b[] += 1
           b[] -= 1
           b[] += 2
           b[]
       end
f (generic function with 1 method)

julia> @code_llvm f(2)
```
```llvm
define i64 @julia_f_63068(i64) #0 !dbg !5 {
top:
  %1 = add i64 %0, 2
  ret i64 %1
}
```

This takes advantage of the LLVM optimizations to get more precise esape info but it doesn't replace allocation elimination in typeinf, which can also split allocations and do more fancy transformations.

The code generated is not 100% optimum (issue from LLVM optimization order) though that might be better when we can run this later in the pipeline. The current placement for 5.0 is arbitrary....

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

### CODE REVIEW

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/22684):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`make -j3`, ProcessExited(2)) [2]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/c18d6a99667aa5183c4d88d6ba4e1093969eff4d/825bfde_vs_bf83397/report.md)
cc @jrevels

**Comment by yuyichao** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/22684#issuecomment-312994011):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`/home/nanosoldier/workdir/tmpSVWZ30/julia -e Pkg.update()`, ProcessSignaled(11)) [0]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/de25fac12d2b5c6341cdd7ad1a52ef29ccf01b99/6b62cff_vs_bf83397/report.md)
cc @jrevels

**Comment by yuyichao** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/22684#issuecomment-313002056):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`/home/nanosoldier/workdir/tmpSzyGHq/julia -e Pkg.update()`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2f39637b7f440e3b9bf3774df392426a6a1a450b/ee76106_vs_bf83397/report.md)
cc @jrevels

**Comment by yuyichao** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by Keno** [comment]
Stealing my thunder, eh? That's ok :). I'll try to review this tomorrow.

File: src/llvm-alloc-opt.cpp [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+}
+
+bool AllocOpt::checkInst(Instruction *I, Instruction *parent, size_t offset, bool &ignore_tag)
+{
+    if (isa<LoadInst>(I))
+        return true;
+    if (auto call = dyn_cast<CallInst>(I)) {
+        if (ptr_from_objref && ptr_from_objref == call->getCalledFunction())
+            return true;
+        // Only use in argument counts, uses in operand bundle doesn't since it cannot escape.
```
Comment (tkelman) [inline src/llvm-alloc-opt.cpp]: doesn't what?

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22684#issuecomment-313144842) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b942172790f82e58f2497f01e1d6fbdfec6a48a6/6c00ebb_vs_bf83397/report.md). cc @jrevels

File: src/llvm-alloc-opt.cpp [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+        }
+    }
+
+    auto &entry = F.getEntryBlock();
+    auto first = &entry.front();
+    for (auto it: allocs) {
+        bool ignore_tag = true;
+        auto orig = it.first;
+        if (optimize && checkUses(orig, 0, ignore_tag)) {
+            // The allocation does not escape or be used in a phi node so none of the derived
```
Comment (tkelman) [inline src/llvm-alloc-opt.cpp]: or get used ?

**Comment by tkelman** [comment]
Does this mean we can finally deprecate `&` in ccall?

File: src/llvm-alloc-opt.cpp (hunk 1 above)
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: Doesn't count.

Comment (tkelman) [inline src/llvm-alloc-opt.cpp]: ah then "uses ... don't" or "use ... doesn't"

**Comment by yuyichao** [comment]
> Stealing my thunder, eh? That's ok :)

FWIW, the main part of this is just ~4hrs of work after I suddenly got interested in checking if the gcframe placement pass is late enough for us to do this. So nothing much would be lost if you already have a better version working ;-p (Plus I've been implementing Jameson's ideas all along  :trollface: )

> Does this mean we can finally deprecate `&` in ccall?

Yes, that should be the case this PR is the best at doing. There are still rare cases where the `&` in `ccall` can be used to reduce allocation but that only happen with a custom `ptr_arg_cconvert` and `ptr_arg_unsafe_convert` and I don't think anyone is doing it. I think master will be open for PR's replacing `&`s with `Ref`s after this is merged and we can deprecate `&` after making sure nothing goes wrong.

File: src/cgutils.cpp [hunk 3]
```diff
@@ -2156,25 +2156,10 @@ static Value *emit_allocobj(jl_codectx_t &ctx, size_t static_size, Value *jt)
 {
     JL_FEAT_REQUIRE(ctx, dynamic_alloc);
     JL_FEAT_REQUIRE(ctx, runtime);
```
Comment (maleadt) [inline src/cgutils.cpp]: Should probably postpone these checks to the actual lowering (but we don't have a `cgctx` there).

Comment (yuyichao) [inline src/cgutils.cpp]: That can be passed as a paramter. What would be the correct behavior if the test fail though?

Comment (maleadt) [inline src/cgutils.cpp]: abort compilation. currently `JL_FEAT_REQUIRE` just calls `jl_error`, and we use the `cgctx` for current function name & line number, so we'd need to do something else anyway (use DebugLoc, I assume). better leave that to a different PR, for now maybe just:

```
if (!JL_FEAT_TEST(ctx, dynamic_alloc)) jl_error(...)
```

Comment (yuyichao) [inline src/cgutils.cpp]: I mean, I don't think we can throw an error in the llvm pass.

Comment (maleadt) [inline src/cgutils.cpp]: Hmm I hadn't considered that, as it worked with `_dump_function` but seems to mess up the JIT indeed. Then again, I only use that feature through `_dump_function`...
Maybe a "used features mask" to be checked after lowering? Let's just leave it at what it is now and put that in a different PR.

File: src/codegen.cpp [hunk 4]
```diff
@@ -6631,24 +6630,14 @@ static void init_julia_llvm_env(Module *m)
...
-                         Function::ExternalLinkage,
-                         "jl_gc_big_alloc", m);
-    add_named_global(jlalloc_big_func, &jl_gc_big_alloc);
+    std::vector<Type*> gc_alloc_args(0);
+    gc_alloc_args.push_back(T_pint8);
+    gc_alloc_args.push_back(T_size);
+    gc_alloc_args.push_back(T_prjlvalue);
+    jl_alloc_obj_func = Function::Create(FunctionType::get(T_prjlvalue, gc_alloc_args, false),
+                                         Function::ExternalLinkage,
+                                         "julia.gc_alloc_obj");
```
Comment (Keno) [inline src/codegen.cpp]: Ideally, this function should get a `noalias` attribute.

**Comment by Keno** [comment]
Couple of suggestions:
1. Use LLVM's `PtrUseVisitor` to avoid excessive stack growth in the recursive approach
2. You can call mem2reg as a cleanup step: http://llvm.org/doxygen/namespacellvm.html#a033a44177ba94b77622aae61ff4fb4b2

**Comment by yuyichao** [comment]
1. `noalias` attribute added.
2. I can't figure how to use `PtrUseVisitor` to handle `AddrSpaceCastInst` and it also seems hard to use it to handle the mutation recursion so I decided to use the pattern I'm more familiar with and used a manual stack similar to the one in the GC.
3. `mem2reg` actually doesn't handle this case (at all). `instcombine` and `sroa` does and the new placement for 5.0 takes advantage of that.
4. Added llvm lifetime intrinsics to reuse stack space. Not sure if there's a LLVM helper function that can make doing that easier...

**Comment by yuyichao** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/22684#issuecomment-313867043):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`make -j3`, ProcessExited(2)) [2]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/daece5ccae64223bff867186d044ec2a326c09e2/c176fc8_vs_0c9a858/report.md)
cc @jrevels

**Comment by yuyichao** [comment]
The issue in the lifetime computation should be fixed...

@nanosoldier runbenchmarks(ALL, vs = ":master")

**Comment by yuyichao** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22684#issuecomment-313870411) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/a376a228adcfa4a551d23cc7062709745bea25db/c7f3ce2_vs_0c9a858/report.md). cc @jrevels

File: src/llvm-alloc-opt.cpp [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+    // to record that.
+    // Iterate through the first dead and find the first safepoint following each of them.
+    while (!first_dead.empty()) {
+        auto I = first_dead.back();
+        first_dead.pop_back();
+        auto bb = I->getParent();
+        if (!bbs.insert(bb).second)
+            continue;
+        if (I == &*bb->begin()) {
+            // The there's no use in or after this bb. If this bb is not dominated by
```
Comment (tkelman) [inline src/llvm-alloc-opt.cpp]: redundant "The" ?

**Comment by vtjnash** [comment]
Need to make sure that for:
```julia
julia> f(fptr, x) = ccall(fptr, Void, (Ref{Integer},), x)
f (generic function with 1 method)

julia> @code_llvm f(C_NULL, 0)
```
We are still emitting the gc-root for `x` even when we elide the allocation of the `Ref` that we used to declare it.

File: base/math.jl [hunk 6]
```diff
@@ -754,9 +754,9 @@ function ieee754_rem_pio2(x::Float64)
     # this is just wrapping up
     # https://github.com/JuliaLang/openspecfun/blob/master/rem_pio2/e_rem_pio2.c
 
-    y = [0.0,0.0]
-    n = ccall((:__ieee754_rem_pio2, openspecfun), Cint, (Float64,Ptr{Float64}), x, y)
-    return (n,y)
+    y = Ref{NTuple{2,Float64}}()
+    n = ccall((:__ieee754_rem_pio2, openspecfun), Cint, (Float64, Ptr{Void}), x, y)
+    return (n, y[])
 end
```
Comment (yuyichao) [inline base/math.jl]: The ordering of the pass on 3.9 and 4.0 still make it slightly less efficient than the llvmcall version.

File: src/cgutils.cpp [hunk 7]
```diff
@@ -2149,25 +2149,10 @@ static Value *emit_allocobj(jl_codectx_t &ctx, size_t static_size, Value *jt)
...
-        Value *pool_offs = ConstantInt::get(T_int32, offset);
-        Value *args[] = {ptls_ptr, pool_offs, ConstantInt::get(T_int32, osize)};
-        v = ctx.builder.CreateCall(prepare_call(jlalloc_pool_func),
-                               ArrayRef<Value*>(args, 3));
-    }
-    tbaa_decorate(tbaa_tag, ctx.builder.CreateStore(maybe_decay_untracked(jt), emit_typeptr_addr(ctx, v)));
-    return v;
+    return ctx.builder.CreateCall(prepare_call(jl_alloc_obj_func),
+                                  {ptls_ptr, ConstantInt::get(T_size, static_size),
+                                          maybe_decay_untracked(jt)});
```
Comment (yuyichao) [inline src/cgutils.cpp]: Does this have to be marked on the callsite?

File: src/llvm-alloc-opt.cpp [hunk 8]
```diff
@@ -0,0 +1,733 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#define DEBUG_TYPE "alloc_opt"
+#undef DEBUG
```
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: .... No idea....

File: src/llvm-alloc-opt.cpp [hunk 9]
```diff
@@ -0,0 +1,733 @@
...
+        if (simple_replace(orig_i, new_i))
+            return;
+        stack.emplace_back(orig_i, new_i);
+        cur = &stack.back();
+    };
+    auto replace_inst = [&] (Instruction *user) {
+        Instruction *orig_i = cur->orig_i;
+        Instruction *new_i = cur->new_i;
+        if (isa<LoadInst>(user) || isa<StoreInst>(user)) {
+            user->replaceUsesOfWith(orig_i, new_i);
```
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: The type of `orig_i` and `new_i` are always the same ignoring addrspace. If they are not, they only differ in addrspace.

File: src/llvm-alloc-opt.cpp [hunk 10]
```diff
@@ -0,0 +1,733 @@
...
+            if (gep->isInBounds()) {
+                new_gep = GetElementPtrInst::CreateInBounds(gep->getSourceElementType(),
+                                                            new_i, IdxOperands,
+                                                            gep->getName(), gep);
+            }
+            else {
+                new_gep = GetElementPtrInst::Create(gep->getSourceElementType(),
+                                                    new_i, IdxOperands,
+                                                    gep->getName(), gep);
+            }
```
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: `copyMetadata` does that.

File: src/llvm-alloc-opt.cpp [hunk 11]
```diff
@@ -0,0 +1,733 @@
...
+        auto pool_offs = ConstantInt::get(T_int32, offset);
+        auto pool_osize = ConstantInt::get(T_int32, osize);
+        newI = CallInst::Create(pool_alloc, {ptls, pool_offs, pool_osize}, None, "", I);
+    }
+    auto tag = I->getArgOperand(2);
+    copyMetadata(newI, I);
+    const auto &dbg = I->getDebugLoc();
+    auto derived = new AddrSpaceCastInst(newI, T_pjlvalue_der, "", I);
+    derived->setDebugLoc(dbg);
+    auto cast = new BitCastInst(derived, T_ppjlvalue_der, "", I);
```
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: AFAICT it's still pretty pedantic....

File: src/llvm-alloc-opt.cpp [hunk 12]
```diff
@@ -0,0 +1,733 @@
...
+            auto cast = new BitCastInst(ptr, T_pjlvalue0, "", first);
+            cast->setDebugLoc(dbg);
+            // Someone might be reading the tag, initialize it.
+            if (!ignore_tag) {
+                auto tag = orig->getArgOperand(2);
+                auto cast2 = new BitCastInst(ptr, T_ppjlvalue0, "", orig);
+                cast2->setDebugLoc(dbg);
+                auto tagaddr = GetElementPtrInst::Create(T_pjlvalue, cast,
+                                                         {ConstantInt::get(T_size, -1)},
+                                                         "", orig);
```
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: Yeah, It was tested but not after making it conditional....... I'm actually not sure how to actually trigger this....

**Comment by yuyichao** [comment]
> `Ref{Integer}`

Being exposed by, fixed in and tested in https://github.com/JuliaLang/julia/pull/22734

File: src/llvm-alloc-opt.cpp [hunk 13]
```diff
@@ -0,0 +1,733 @@
...
+}
+
+bool AllocOpt::checkInst(Instruction *I, CheckInstStack &stack, std::set<Instruction*> &uses,
+                         bool &ignore_tag)
+{
+    uses.clear();
+    if (I->user_empty())
+        return true;
+    CheckInstFrame cur{I, 0, I->user_begin(), I->user_end()};
+    stack.clear();
```
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: This one might actually not be empty due to the short circuit return. (The other two should.)

File: src/llvm-alloc-opt.cpp [hunk 14]
```diff
@@ -0,0 +1,746 @@
...
+    dest->setDebugLoc(src->getDebugLoc());
+#else
+    dest->copyMetadata(*src);
+#endif
+}
+
+/**
+ * Promote `julia.gc_alloc_obj` which do not have escaping root to a alloca and
+ * lower other ones to real GC allocation.
+ * Uses that are not considerer to escape the object (i.e. heap address) includes,
```
Comment (tkelman) [inline src/llvm-alloc-opt.cpp]: considered ?

File: src/llvm-alloc-opt.cpp [hunk 15]
```diff
@@ -0,0 +1,746 @@
...
+ *
+ * * load
+ * * `pointer_from_objref`
+ * * `ccall` gcroot array (`jl_roots` operand bundle)
+ * * store (as address)
+ * * addrspacecast, bitcast, getelementptr
+ *
+ *     The results of these cast instructions will be scanned recursively.
+ *
+ * All other uses are considered escape conservatively.
```
Comment (tkelman) [inline src/llvm-alloc-opt.cpp]: considered to escape ?

File: src/llvm-alloc-opt.cpp [hunk 16]
```diff
@@ -0,0 +1,746 @@
...
+}
+
+void AllocOpt::LifetimeMarker::insertEnd(Instruction *ptr, Constant *sz, Instruction *insert)
+{
+    BasicBlock::iterator it(insert);
+    BasicBlock::iterator begin(insert->getParent()->begin());
+    // Makes sure that the end is inserted before nearby start.
+    // We insert start before the allocation call, if it is the first safepoint we find for
+    // another instruction, it's better if we insert the end before the start instead of the
+    // allocation so that the two allocation do not have overlapping lifetime.
```
Comment (tkelman) [inline src/llvm-alloc-opt.cpp]: so that the two allocations ?

File: src/llvm-alloc-opt.cpp [hunk 17]
```diff
@@ -0,0 +1,746 @@
...
+    auto check_inst = [&] (Instruction *inst) {
+        if (isa<LoadInst>(inst))
+            return true;
+        if (auto call = dyn_cast<CallInst>(inst)) {
+            // TODO: on LLVM 5.0 we may need to handle certain llvm intrinsics
+            // including `memcpy`, `memset` etc. We might also need to handle
+            // `memcmp` by coverting to our own intrinsic and lower it after the gc root pass.
+            if (ptr_from_objref && ptr_from_objref == call->getCalledFunction())
+                return true;
+            // Only use in argument counts, use in operand bundle doesn't since it cannot escape.
```
Comment (Keno) [inline src/llvm-alloc-opt.cpp]: Only skip the "jl_roots" arg bundle? In general arg bundles escape in unknown ways per the spec.

**Comment by Keno** [comment]
I'd like to see some LLVM level tests. Also, why introduce a new allocation intrinsic, rather than operating on the actual GC functions? It seems like it may be useful to put this pass in the pipeline twice (e.g. if we want to do LLVM level inlining).

File: src/llvm-alloc-opt.cpp (hunk 17 above)
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: It is only skipping `jl_roots` arg bundle. I'll update the comment.

**Comment by yuyichao** [comment]
> why introduce a new allocation intrinsic

1. The `jl_gc_pool_alloc` does not provide the actually needed size (`osize` contains tag and is rounded up by a lot, especially for small objects like `Int8`, `Float32`)
2. It avoids introducing the tag store in the IR too early. That is much harder to optimize out.

I've also thought about running this twice. It should be easy to do by disabling the intrinsic lowering with a flag.

**Comment by yuyichao** [comment]
All actionable review comments addressed.

**Comment by Keno** [comment]
Would it make sense to do the lowering from intrinsics to actual call as part of GC root placement? That pass runs only once anyway and it doesn't make sense to do any allocation optimization afterward.

File: test/llvmpasses/alloc-opt.jl [hunk 18]
```diff
@@ -0,0 +1,204 @@
+# RUN: julia --startup-file=no %s | opt -load libjulia.so -AllocOpt -S - | FileCheck %s
+
+## WIP
+
+isz = sizeof(UInt) == 8 ? "i64" : "i32"
```
Comment (Keno) [inline test/llvmpasses/alloc-opt.jl]: We should really not have the pass depend on the how architecture (and have it look at the target arch/triple instead).

**Comment by yuyichao** [comment]
> as part of GC root placement

I think it makes less sense than doing it here. This pass already need to collect all the allocations and the last invokation of this pass might as well lower the remaining ones.

**Comment by Keno** [comment]
Why is the lowering not relatively trivial?

**Comment by Keno** [comment]
Or if it is, GC root lowering looks at all the calls and rewrites most of them anyway (because of the cc convention), so it seems like a fine place to do it.

**Comment by Keno** [comment]
The concern is that this pass might not know whether it's the last one (in IPO pipelines the pass manager will automatically rerun parts of the pipeline).

File: test/llvmpasses/alloc-opt.jl (hunk 18 above)
Comment (yuyichao) [inline test/llvmpasses/alloc-opt.jl]: While the integer size is easy to fix, its actually a pretty complex logic. The pass relies on details on object layout which not only depend on 32/64bit but also the architecture and even the OS. In particular, the pool offset depends on other members in the tls buffer including the size of `jl_jmp_buf`.

P.S. just notice that the `WIP` should be removed.....

**Comment by yuyichao** [comment]
Moved the intrinsic lowering to gcframe lowering pass and updated the test.

**Comment by yuyichao** [comment]
Travis failure looks unrelated and is happenning everywhere....

**Comment by yuyichao** [comment]
More comments?

**Comment by Keno** [comment]
I'll try to take another pass through this tonight or tomorrow morning.

File: src/llvm-alloc-opt.cpp [hunk 19]
```diff
@@ -0,0 +1,667 @@
...
+            finish_cur();
+            if (stack.empty())
+                return;
+            cur = stack.back();
+            stack.pop_back();
+        }
+    }
+}
+
+bool AllocOpt::isSafepoint(Instruction *inst)
```
Comment (Keno) [inline src/llvm-alloc-opt.cpp]: Pull this out into a helper for both this and the GC lowering code?

File: src/llvm-alloc-opt.cpp [hunk 20]
```diff
@@ -0,0 +1,667 @@
...
+        }
+    }
+    return true;
+}
+
+bool AllocOpt::runOnFunction(Function &F)
+{
+    if (!alloc_obj)
+        return false;
+    std::map<CallInst*,size_t> allocs;
```
Comment (Keno) [inline src/llvm-alloc-opt.cpp]: std::map is generally discouraged in LLVM passes because of nondeterministic iteration order. Can you use a data structure with deterministic iteration?

File: src/llvm-alloc-opt.cpp [hunk 21]
```diff
@@ -0,0 +1,667 @@
...
+        if (isa<LoadInst>(inst))
+            return true;
+        if (auto call = dyn_cast<CallInst>(inst)) {
+            // TODO: on LLVM 5.0 we may need to handle certain llvm intrinsics
+            // including `memcpy`, `memset` etc. We might also need to handle
+            // `memcmp` by coverting to our own intrinsic and lower it after the gc root pass.
+            if (ptr_from_objref && ptr_from_objref == call->getCalledFunction())
+                return true;
+            auto opno = use->getOperandNo();
+            // Uses in `jl_roots` operand bundle are not counted as escaping, everything else do.
```
Comment (Keno) [inline src/llvm-alloc-opt.cpp]: Nit: `everything else is`.

File: src/llvm-alloc-opt.cpp [hunk 22]
```diff
@@ -0,0 +1,667 @@
...
+    };
+    auto push_frame = [&] (Instruction *orig_i, Instruction *new_i) {
+        if (simple_replace(orig_i, new_i))
+            return;
+        stack.push_back(cur);
+        cur = {orig_i, new_i};
+    };
+    // Both `orig_i` and `new_i` should be pointer of the same type
+    // but possibly different address spaces. `new_i` is always in addrspace 0.
+    auto replace_inst = [&] (Instruction *user) {
```
Comment (Keno) [inline src/llvm-alloc-opt.cpp]: You can technically mutate instructions in place, but I'm fine with this implementation as well.

File: src/llvm-alloc-opt.cpp (hunk 19 above)
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: Does the lowering pass actually need this function? It has similar logic but does different things for different branches.

File: src/llvm-alloc-opt.cpp (hunk 20 above)
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: Seems that I actually don't need to lookup anything in it so I'll just use a vector instead...

File: src/llvm-alloc-opt.cpp (hunk 19 above)
Comment (Keno) [inline src/llvm-alloc-opt.cpp]: It doesn't need this function, but I'd like to at least share the "Known functions emitted in codegen that are not safepoints" part, so we don't have to make changes there in multiple places.

**Comment by yuyichao** [comment]
Updated comment and switched to a vector instead for recording the allocations.
I did not pull the `isSafePoint` out since there currently isn't a use in the lowering pass. We can certainly split it out when there is one.

File: src/llvm-alloc-opt.cpp (hunk 19 above)
Comment (yuyichao) [inline src/llvm-alloc-opt.cpp]: Also note that when the lowering pass recognize more functions as non safepoint, we don't necessarily want to update this pass to include those. Here I included the same list just to be safe. In principle, this needs a list of functions that codegen assums to be not safepoint. There can be functions that are never safepoint but as long as neither codegen or llvm can insert a call into a unsafe use chain it doesn't need to be treated as non safepoint here.

**Comment by yuyichao** [comment]
I'd like to not make any other changes about the organization, especially since I still don't think it's a good idea to move the allocation that late in the pipeline. It disables the llvm constant folding of write barrier on 5.0....

There are two FreeBSD timeout and I have no way to debug it so I'll just wait for someone else to figure out what's wrong there or fix it or merge this PR.

**Comment by yuyichao** [comment]
And for anyone who want to use this on master, the hanging test on FreeBSD is `file`.

**Comment by iblis17** [comment]
the `distributed` testsuit absent in log also.
FreeBSD CI keep hanging randomly ... It seems start from 5ea8c7c8f6
I guess that the test case add in #22566 another bug...

**Comment by yuyichao** [comment]
> the distributed testsuit absent in log also.
JULIA_TEST_MAX_RSS is set so it is moved to worker one and will only run if everything else finishes.

**Comment by iblis17** [comment]
👌 
In case of hanging `file`, feel free to rerun CI.
There is a `rebuild` button at the top-right corner of buildbot UI.

**Comment by yuyichao** [comment]
[Restarted one](https://julia.iblis.cnmc.tw/#/builders/1/builds/577) passed so I'm merging this now....

**Comment by iblis17** [comment]
got some compliation warnning from master e1a604e088
```cpp
/usr/home/iblis/git/julia/src/llvm-alloc-opt.cpp:639:53: warning: braces around scalar initializer [-Wbraced-scalar-init]                         
                                                    {ConstantInt::get(T_size, -1)});                                                              
                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                
/usr/home/iblis/git/julia/src/llvm-late-gc-lowering.cpp:1173:50: warning: braces around scalar initializer [-Wbraced-scalar-init]                 
                                                 {ConstantInt::get(T_size, -1)});                                                                 
                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                   
1 warning generated.                                                                                                                              
1 warning generated.                                                                                                                              
```

**Comment by JeffBezanson** [comment]
This is awesome!

From the nanosoldier run, it looks like we don't have any benchmarks that meaningfully improve with this? We should try to add some.

**Comment by yuyichao** [comment]
Regression on this will probably show up in future benchmarks as we have more and more use of `Ref` in `ccall`s.

We are in general very good at avoiding performance pitfails in the benchmarks....
